core/finality-grandpa/src/communication/mod.rs

```diff
@@ -363,71 +371,41 @@ pub(crate) fn check_message_sig<Block: BlockT>(
 	}
 }
 
-/// Whether we've voted already during a prior run of the program.
-#[derive(Decode, Encode)]
-pub(crate) enum HasVoted {
-	/// Has not voted already in this round.
-	#[codec(index = "0")]
-	No,
-	/// Has cast a proposal.
-	#[codec(index = "1")]
-	Proposed,
-	/// Has cast a prevote.
-	#[codec(index = "2")]
-	Prevoted,
-	/// Has cast a precommit (implies prevote.)
-	#[codec(index = "3")]
-	Precommitted,
-}
-
-impl HasVoted {
-	#[allow(unused)]
-	fn can_propose(&self) -> bool {
-		match *self {
-			HasVoted::No => true,
-			HasVoted::Proposed | HasVoted::Prevoted | HasVoted::Precommitted => false,
-		}
-	}
-
-	fn can_prevote(&self) -> bool {
-		match *self {
-			HasVoted::No | HasVoted::Proposed => true,
-			HasVoted::Prevoted | HasVoted::Precommitted => false,
-		}
-	}
-
-	fn can_precommit(&self) -> bool {
-		match *self {
-			HasVoted::No | HasVoted::Proposed | HasVoted::Prevoted => true,
-			HasVoted::Precommitted => false,
-		}
-	}
-}
-
-/// A sink for outgoing messages to the network.
+/// A sink for outgoing messages to the network. Any messages that are sent will
+/// be replaced, as appropriate, according to the given `HasVoted`.
 struct OutgoingMessages<Block: BlockT, N: Network<Block>> {
 	round: u64,
 	set_id: u64,
 	locals: Option<(Arc<ed25519::Pair>, AuthorityId)>,
 	sender: mpsc::UnboundedSender<SignedMessage<Block>>,
 	network: N,
-	has_voted: HasVoted,
+	has_voted: HasVoted<Block>,
 }
 
 impl<Block: BlockT, N: Network<Block>> Sink for OutgoingMessages<Block, N>
 {
 	type SinkItem = Message<Block>;
 	type SinkError = Error;
 
-	fn start_send(&mut self, msg: Message<Block>) -> StartSend<Message<Block>, Error> {
-		// only sign if we haven't voted in this round already.
-		let should_sign = match msg {
-			grandpa::Message::Prevote(_) => self.has_voted.can_prevote(),
-			grandpa::Message::Precommit(_) => self.has_voted.can_precommit(),
-		};
+	fn start_send(&mut self, mut msg: Message<Block>) -> StartSend<Message<Block>, Error> {
+		// if we've voted on this round previously under the same key, send that vote instead
+		match &mut msg {
+			grandpa::Message::PrimaryPropose(ref mut vote) =>
+				if let Some(propose) = self.has_voted.propose() {
+					*vote = propose;
```

Comment:
GRANDPA treats the same votes with different signatures as equivocations. ed25519 signatures would end up being the same for all votes, (I think for BLS too) but perhaps we want to store the votes' signatures in the DB as well.